Let the compiler generate more comparison operators in WebCore

cdumez opened the pull request:
#### 32a30dfe6ffb0b6380bca678439c115b60da5be9
<pre>
Let the compiler generate more comparison operators in WebCore
<a href="https://bugs.webkit.org/show_bug.cgi?id=260943">https://bugs.webkit.org/show_bug.cgi?id=260943</a>

Reviewed by Timothy Hatcher, Ryosuke Niwa and Darin Adler.

Let the compiler generate more comparison operators in WebCore now that we
support C++20.

* Source/WebCore/Modules/highlight/AppHighlightRangeData.h:
(WebCore::AppHighlightRangeData::NodePathComponent::operator== const): Deleted.
* Source/WebCore/Modules/indexeddb/IDBDatabaseIdentifier.h:
(WebCore::IDBDatabaseIdentifier::operator== const): Deleted.
* Source/WebCore/Modules/indexeddb/shared/IDBResourceIdentifier.h:
(WebCore::IDBResourceIdentifier::operator== const): Deleted.
* Source/WebCore/Modules/mediasession/MediaPositionState.h:
(WebCore::MediaPositionState::operator== const): Deleted.
* Source/WebCore/Modules/permissions/PermissionDescriptor.h:
(WebCore::PermissionDescriptor::operator== const): Deleted.
* Source/WebCore/Modules/webaudio/BaseAudioContext.h:
* Source/WebCore/PAL/pal/SessionID.h:
(PAL::SessionID::operator== const): Deleted.
* Source/WebCore/contentextensions/ContentExtensionActions.cpp:
(WebCore::ContentExtensions::ModifyHeadersAction::operator== const): Deleted.
(WebCore::ContentExtensions::ModifyHeadersAction::ModifyHeaderInfo::operator== const): Deleted.
(WebCore::ContentExtensions::RedirectAction::operator== const): Deleted.
(WebCore::ContentExtensions::RedirectAction::URLTransformAction::operator== const): Deleted.
(WebCore::ContentExtensions::RedirectAction::URLTransformAction::QueryTransform::operator== const): Deleted.
(WebCore::ContentExtensions::RedirectAction::URLTransformAction::QueryTransform::QueryKeyValue::operator== const): Deleted.
* Source/WebCore/contentextensions/ContentExtensionActions.h:
(WebCore::ContentExtensions::ActionWithoutMetadata::operator== const): Deleted.
(WebCore::ContentExtensions::ActionWithStringMetadata::operator== const): Deleted.
* Source/WebCore/contentextensions/ContentExtensionRule.h:
(WebCore::ContentExtensions::Trigger::operator== const): Deleted.
(WebCore::ContentExtensions::Action::operator== const): Deleted.
(WebCore::ContentExtensions::ContentExtensionRule::operator== const): Deleted.
* Source/WebCore/contentextensions/NFAToDFA.cpp:
* Source/WebCore/contentextensions/Term.h:
(WebCore::ContentExtensions::Term::CharacterSet::operator== const): Deleted.
(WebCore::ContentExtensions::Term::Group::operator== const): Deleted.
* Source/WebCore/css/CSSCounterStyleDescriptors.h:
(WebCore::CSSCounterStyleDescriptors::operator== const):
(WebCore::CSSCounterStyleDescriptors::Symbol::operator== const): Deleted.
(WebCore::CSSCounterStyleDescriptors::Pad::operator== const): Deleted.
(WebCore::CSSCounterStyleDescriptors::NegativeSymbols::operator== const): Deleted.
* Source/WebCore/css/CSSCounterStyleRegistry.cpp:
(WebCore::CSSCounterStyleRegistry::operator== const):
* Source/WebCore/css/CSSCustomPropertyValue.h:
* Source/WebCore/css/CSSFontFace.h:
* Source/WebCore/css/CSSGradientValue.cpp:
(WebCore::CSSLinearGradientValue::Angle::operator==):
(WebCore::CSSPrefixedLinearGradientValue::operator==):
(WebCore::CSSRadialGradientValue::operator==):
(WebCore::CSSPrefixedRadialGradientValue::operator==):
(WebCore::CSSConicGradientValue::operator==):
* Source/WebCore/css/CSSGradientValue.h:
(WebCore::CSSGradientColorInterpolationMethod::legacyMethod):
(WebCore::operator==):

Canonical link: <a href="https://commits.webkit.org/267512@main">https://commits.webkit.org/267512@main</a>
</pre>
<!--EWS-Status-Bubble-Start-->
https://github.com/WebKit/WebKit/commit/1b89dcf61ab25cde0a567c60498a976ab89d572a

| Misc | iOS, tvOS & watchOS  | macOS  | Linux |  Windows |
| ----- | ---------------------- | ------- |  ----- |  --------- |
| [✅ 🧪 style](https://ews-build.webkit.org/#/builders/38/builds/16804 "Passed style check") | [✅ 🛠 ios](https://ews-build.webkit.org/#/builders/26/builds/17127 "Built successfully") | [✅ 🛠 mac](https://ews-build.webkit.org/#/builders/14/builds/17581 "Built successfully") | [✅ 🛠 wpe](https://ews-build.webkit.org/#/builders/5/builds/18586 "Built successfully") | [✅ 🛠 wincairo](https://ews-build.webkit.org/#/builders/32/builds/15741 "Built successfully") 
| [✅ 🧪 bindings](https://ews-build.webkit.org/#/builders/9/builds/16994 "Passed tests") | [✅ 🛠 ios-sim](https://ews-build.webkit.org/#/builders/23/builds/20360 "Built successfully") | [✅ 🛠 mac-AS-debug](https://ews-build.webkit.org/#/builders/16/builds/17266 "Built successfully") | [  ~~🧪 wpe-wk2~~](https://ews-build.webkit.org/#/builders/34/builds/18018 "The change is no longer eligible for processing. Pull Request was already closed when EWS attempted to process it.") | 
| [✅ 🧪 webkitperl](https://ews-build.webkit.org/#/builders/11/builds/17003 "Passed tests") | [✅ 🧪 ios-wk2](https://ews-build.webkit.org/#/builders/15/builds/17368 "Passed tests") | [✅ 🧪 api-mac](https://ews-build.webkit.org/#/builders/18/builds/14550 "Passed tests") | [✅ 🛠 gtk](https://ews-build.webkit.org/#/builders/2/builds/19385 "Built successfully") | 
| | [✅ 🧪 ios-wk2-wpt](https://ews-build.webkit.org/#/builders/39/builds/14612 "Passed tests") | [✅ 🧪 mac-wk1](https://ews-build.webkit.org/#/builders/10/builds/15234 "Passed tests") | [  ~~🧪 gtk-wk2~~](https://ews-build.webkit.org/#/builders/1/builds/21980 "The change is no longer eligible for processing. Commit was outdated when EWS attempted to process it.") | 
| | [✅ 🧪 api-ios](https://ews-build.webkit.org/#/builders/13/builds/15601 "Passed tests") | [✅ 🧪 mac-wk2](https://ews-build.webkit.org/#/builders/36/builds/15401 "Passed tests") | [✅ 🧪 api-gtk](https://ews-build.webkit.org/#/builders/21/builds/19714 "Passed tests") | 
| | [✅ 🛠 tv](https://ews-build.webkit.org/#/builders/7/builds/16004 "Built successfully") | [✅ 🧪 mac-AS-debug-wk2](https://ews-build.webkit.org/#/builders/17/builds/13572 "Passed tests") | | 
| | [✅ 🛠 tv-sim](https://ews-build.webkit.org/#/builders/30/builds/15179 "Built successfully") | | | 
| [✅ 🛠 🧪 merge](https://ews-build.webkit.org/#/builders/19/builds/4026 "Built successfully and passed tests") | [✅ 🛠 watch](https://ews-build.webkit.org/#/builders/4/builds/19543 "Built successfully") | | | 
| | [✅ 🛠 watch-sim](https://ews-build.webkit.org/#/builders/31/builds/15835 "Built successfully") | | | 
<!--EWS-Status-Bubble-End-->

darinadler commented on Source/WebCore/css/CSSGradientValue.h:
```diff
@@ -232,38 +238,48 @@ class CSSRadialGradientValue final : public CSSValue {
         Ref<CSSPrimitiveValue> length; // <length [0,∞]>
         std::optional<CSSGradientPosition> position;
+
+        friend bool operator==(const Length& a, const Length&);
```
There’s a stray "a" here.

cdumez replied:
Fixed.

darinadler commented on Source/WebCore/css/CSSGradientValue.h:
```diff
@@ -232,38 +238,48 @@ class CSSRadialGradientValue final : public CSSValue {
     struct Size {
         std::pair<Ref<CSSPrimitiveValue>, Ref<CSSPrimitiveValue>> size; // <length-percentage [0,∞]>, <length-percentage [0,∞]>
         std::optional<CSSGradientPosition> position;
+        friend bool operator==(const Size& a, const Size&);
```
There’s a stray "a" here.

darinadler commented on Source/WebCore/css/CSSGradientValue.h:
```diff
@@ -232,38 +238,48 @@ class CSSRadialGradientValue final : public CSSValue {
     struct EllipseOfSize {
         std::pair<Ref<CSSPrimitiveValue>, Ref<CSSPrimitiveValue>> size; // <length-percentage [0,∞]>, <length-percentage [0,∞]>
         std::optional<CSSGradientPosition> position;
+        friend bool operator==(const EllipseOfSize&, const EllipseOfSize&);
```
This is missing the `= default`.

cdumez replied:
This is not defaulted, it has an implementation in the cpp. However, I add to add the declaration in the header otherwise some of the defaulted comparison operators were complaining.

The same comment applies to the other places you commented on.

darinadler commented on Source/WebCore/css/CSSGradientValue.h:
```diff
@@ -302,25 +318,27 @@ class CSSRadialGradientValue final : public CSSValue {
     };
     struct MeasuredSize {
         std::pair<Ref<CSSPrimitiveValue>, Ref<CSSPrimitiveValue>> size; // <length-percentage [0,∞]>, <length-percentage [0,∞]>
+        friend bool operator==(const MeasuredSize&, const MeasuredSize&);
```
This is missing the `= default`.

darinadler commented on Source/WebCore/css/CSSGradientValue.h:
```diff
@@ -411,11 +427,16 @@ bool operator==(const CSSDeprecatedRadialGradientValue::Data&, const CSSDeprecat
-    struct Angle { RefPtr<CSSPrimitiveValue> value; };
+    struct Angle {
+        RefPtr<CSSPrimitiveValue> value;
+        friend bool operator==(const Angle&, const Angle&);
```
This is missing the `= default`.

cdumez committed:
Let the compiler generate more comparison operators in WebCore
https://bugs.webkit.org/show_bug.cgi?id=260943

Reviewed by Timothy Hatcher, Ryosuke Niwa and Darin Adler.

Let the compiler generate more comparison operators in WebCore now that we
support C++20.

* Source/WebCore/Modules/highlight/AppHighlightRangeData.h:
(WebCore::AppHighlightRangeData::NodePathComponent::operator== const): Deleted.
* Source/WebCore/Modules/indexeddb/IDBDatabaseIdentifier.h:
(WebCore::IDBDatabaseIdentifier::operator== const): Deleted.
* Source/WebCore/Modules/indexeddb/shared/IDBResourceIdentifier.h:
(WebCore::IDBResourceIdentifier::operator== const): Deleted.
* Source/WebCore/Modules/mediasession/MediaPositionState.h:
(WebCore::MediaPositionState::operator== const): Deleted.
* Source/WebCore/Modules/permissions/PermissionDescriptor.h:
(WebCore::PermissionDescriptor::operator== const): Deleted.
* Source/WebCore/Modules/webaudio/BaseAudioContext.h:
* Source/WebCore/PAL/pal/SessionID.h:
(PAL::SessionID::operator== const): Deleted.
* Source/WebCore/contentextensions/ContentExtensionActions.cpp:
(WebCore::ContentExtensions::ModifyHeadersAction::operator== const): Deleted.
(WebCore::ContentExtensions::ModifyHeadersAction::ModifyHeaderInfo::operator== const): Deleted.
(WebCore::ContentExtensions::RedirectAction::operator== const): Deleted.
(WebCore::ContentExtensions::RedirectAction::URLTransformAction::operator== const): Deleted.
(WebCore::ContentExtensions::RedirectAction::URLTransformAction::QueryTransform::operator== const): Deleted.
(WebCore::ContentExtensions::RedirectAction::URLTransformAction::QueryTransform::QueryKeyValue::operator== const): Deleted.
* Source/WebCore/contentextensions/ContentExtensionActions.h:
(WebCore::ContentExtensions::ActionWithoutMetadata::operator== const): Deleted.
(WebCore::ContentExtensions::ActionWithStringMetadata::operator== const): Deleted.
* Source/WebCore/contentextensions/ContentExtensionRule.h:
(WebCore::ContentExtensions::Trigger::operator== const): Deleted.
(WebCore::ContentExtensions::Action::operator== const): Deleted.
(WebCore::ContentExtensions::ContentExtensionRule::operator== const): Deleted.
* Source/WebCore/contentextensions/NFAToDFA.cpp:
* Source/WebCore/contentextensions/Term.h:
(WebCore::ContentExtensions::Term::CharacterSet::operator== const): Deleted.
(WebCore::ContentExtensions::Term::Group::operator== const): Deleted.
* Source/WebCore/css/CSSCounterStyleDescriptors.h:
(WebCore::CSSCounterStyleDescriptors::operator== const):
(WebCore::CSSCounterStyleDescriptors::Symbol::operator== const): Deleted.
(WebCore::CSSCounterStyleDescriptors::Pad::operator== const): Deleted.
(WebCore::CSSCounterStyleDescriptors::NegativeSymbols::operator== const): Deleted.
* Source/WebCore/css/CSSCounterStyleRegistry.cpp:
(WebCore::CSSCounterStyleRegistry::operator== const):
* Source/WebCore/css/CSSCustomPropertyValue.h:
* Source/WebCore/css/CSSFontFace.h:
* Source/WebCore/css/CSSGradientValue.cpp:
(WebCore::CSSLinearGradientValue::Angle::operator==):
(WebCore::CSSPrefixedLinearGradientValue::operator==):
(WebCore::CSSRadialGradientValue::operator==):
(WebCore::CSSPrefixedRadialGradientValue::operator==):
(WebCore::CSSConicGradientValue::operator==):
* Source/WebCore/css/CSSGradientValue.h:
(WebCore::CSSGradientColorInterpolationMethod::legacyMethod):
(WebCore::operator==):

Canonical link: https://commits.webkit.org/267512@main
